### PR TITLE
Fixed #30215 -- Fixed autoreloader crash for modules without __spec__.

### PR DESCRIPTION
**Pull request by shiningfm**
https://code.djangoproject.com/ticket/30215

### CODE REVIEW

File: tests/utils_tests/test_autoreload.py [hunk 1]
```diff
@@ -124,6 +125,11 @@ def test_weakref_in_sys_module(self):
         self.addCleanup(lambda: sys.modules.pop('time_proxy', None))
         list(autoreload.iter_all_python_module_files())  # No crash.
 
+    def test_module_without_spec(self):
+        module = types.ModuleType("test_module")
```
Comment (charettes) [inline tests/utils_tests/test_autoreload.py]: Single quote strings

```suggestion
        module = types.ModuleType('test_module')
```

File: django/utils/autoreload.py [hunk 2]
```diff
@@ -111,7 +111,9 @@ def iter_modules_and_files(modules, extra_files):
         # During debugging (with PyDev) the 'typing.io' and 'typing.re' objects
         # are added to sys.modules, however they are types not modules and so
         # cause issues here.
-        if not isinstance(module, ModuleType) or module.__spec__ is None:
+        if (not isinstance(module, ModuleType) or
+                not hasattr(module, '__spec__') or
+                module.__spec__ is None):
```
Comment (charettes) [inline django/utils/autoreload.py]: `getattr(module, '__spec__', None) is None` is more concise.

File: tests/utils_tests/test_autoreload.py [hunk 3]
```diff
@@ -124,6 +125,11 @@ def test_weakref_in_sys_module(self):
...
+        del module.__spec__
+        autoreload.iter_modules_and_files((module,), frozenset([]))
```
Comment (charettes) [inline tests/utils_tests/test_autoreload.py]: You can pass `set()`, no need to create a list to initialize an empty `frozenset` (or `set`), it be great to assert the return value as well.

File: django/utils/autoreload.py (hunk 2 above)
Comment (charettes) [reply django/utils/autoreload.py]: You can drop the `or module.__spec__ is None`, both are now contained in the `module.__spec__ is None` check.

Comment (shiningfm) [reply django/utils/autoreload.py]: Right, sorry, I should have seen that...

**Pull request merged**